core/domain/collection_services.py

```diff
@@ -256,6 +257,52 @@ def get_collection_titles_and_categories(collection_ids):
     return result
 
 
+def get_completed_exploration_ids(user_id, collection_id):
+    """Returns a list of explorations the user has completed within the context
+    of the provided collection. Returns an empty list if the user has not yet
+    completed any explorations within the collection.
+    """
+    completion_model = user_models.ExplorationInCollectionCompletionModel.get(
+        user_id, collection_id)
+    return completion_model.completed_explorations if completion_model else []
+
+
+def get_next_exploration_ids_to_complete_by_user(user_id, collection_id):
+    """Returns a list of exploration IDs in the specified collection that the
+    given user has not yet attempted and has the prerequisite skills to play.
+
+    Returns the collection's initial explorations if the user has yet to
+    complete any explorations within the collection. Returns an empty list if
+    the user has completed all of the explorations within the collection.
+
+    See collection_domain.Collection.get_next_exploration_ids for more
+    information.
+    """
+    completed_exploration_ids = get_completed_exploration_ids(
+        user_id, collection_id)
+
+    collection = get_collection_by_id(collection_id)
+    if completed_exploration_ids:
+        return collection.get_next_exploration_ids(completed_exploration_ids)
+    else:
+        # The user has yet to complete any explorations inside the collection.
+        return collection.init_exploration_ids
+
+
+def record_played_exploration_in_collection_context(
+        user_id, collection_id, exploration_id):
+    completion_model = (
+        user_models.ExplorationInCollectionCompletionModel.get_or_create(
+            user_id, collection_id))
+
+    completed_explorations = copy.deepcopy(
+        completion_model.completed_explorations)
+    if exploration_id not in completed_explorations:
```

Comment:
Any reason not to do:

    if exploration_id not in completion_model.completed_explorations:
        completion_model.completed_explorations.append(exploration_id)
        completion_model.put()